grammarinator/generate.py

```diff
@@ -126,12 +121,22 @@ def create_new_test(self, index, lock):
             logger.warning('Test generation failed.', exc_info=e)
             return self.create_new_test(index, lock)
 
-        test_fn = self.out_format % index if self.out_format else None
+        test_fn = None
+        if self.out_format and '%d' in self.out_format:
+            test_fn = self.out_format % index
+        else:
+            test_fn = self.out_format
         tree.root = Generator.transform(tree.root, self.transformers)
 
         tree_fn = None
-        if self.population and self.keep_trees:
-            tree_fn = join(self.population.directory, basename(test_fn) + Tree.extension)
+        if self.population and self.keep_trees and self.out_format:
+            tree_basename = basename(self.out_format)
+            if tree_basename and '%d' not in tree_basename:
+                base, ext = splitext(self.out_format)
+                tree_basename = '{base}%d{ext}'.format(base=base, ext=ext) if ext else join(base, '%d')
+            
+            tree_basename = tree_basename % index
+            tree_fn = join(self.population.directory, tree_basename + Tree.extension)
```

Comment:
I guess I wasn't sure how you wanted the trees saved. With this change, if the user specifies a write to only a single test file then the tree file name will also only be a single file name. This makes sense if you want the user to be able to create multiple test files and multiple tree files OR a single test file and single tree file. I'll go ahead and make the change.

Reply:
Keeping the trees makes possible to mutate or recombine trees/tests generated earlier. If we only keep a single tree (always the last one), then this feature will lose its efficacy, since it won't be able to reuse previous results.

Reply:
Right. Your request only keeps the last one though?

Reply:
Ah, you are right. I missed it, too. I was focusing on handling stdout output correctly. Please see my updated review and please squash the commits into a single one and rebase it to the current master.

Reply:
Done